Stabilize the `re_rebalance_coherence` feature

ohadravid opened the pull request:
This PR stabilizes [RFC 2451](https://rust-lang.github.io/rfcs/2451-re-rebalancing-coherence.html), re-rebalance coherence.

Changes include removing the attribute from tests which tested both the old and new behavior, moving the feature to `accepted` and removing the old logic.

I'll also open a [PR](https://github.com/rust-lang-nursery/reference/pull/703) against the reference, updating it with the content of the RFC.

Closes #63599

r? @nikomatsakis

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=12057&view=logs&jobId=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/12057/logs/49), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/12057/logs/49)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-10-27T20:11:21.1486910Z ##[command]git remote add origin https://github.com/rust-lang/rust
2019-10-27T20:11:22.0107032Z ##[command]git config gc.auto 0
2019-10-27T20:11:22.0109882Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2019-10-27T20:11:22.0112165Z ##[command]git config --get-all http.proxy
2019-10-27T20:11:22.0114910Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/65879/merge:refs/remotes/pull/65879/merge
---
2019-10-27T21:12:32.5113397Z .................................................................................................... 1500/9188
2019-10-27T21:12:38.5698607Z .................................................................................................... 1600/9188
2019-10-27T21:12:48.3209189Z ............................................................................................i....... 1700/9188
2019-10-27T21:12:56.6319584Z ........i........................................................................................... 1800/9188
2019-10-27T21:13:03.7194630Z ..................................................................................iiiii............. 1900/9188
2019-10-27T21:13:26.1380647Z .................................................................................................... 2100/9188
2019-10-27T21:13:28.6221798Z .................................................................................................... 2200/9188
2019-10-27T21:13:31.3670291Z .................................................................................................... 2300/9188
2019-10-27T21:13:45.0037102Z .................................................................................................... 2400/9188
---
2019-10-27T21:16:52.1567434Z .....................................................................................i.............. 4700/9188
2019-10-27T21:16:59.8874174Z .i.................................................................................................. 4800/9188
2019-10-27T21:17:10.2855092Z .................................................................................................... 4900/9188
2019-10-27T21:17:16.2000757Z .................................................................................................... 5000/9188
2019-10-27T21:17:28.4316084Z ......................................................................................ii.ii......... 5100/9188
2019-10-27T21:17:32.2964126Z ..i................................................................................................. 5200/9188
2019-10-27T21:17:47.0549575Z .................................................................................................... 5400/9188
2019-10-27T21:17:53.9394706Z ........................................................i........................................... 5500/9188
2019-10-27T21:18:00.9589769Z .................................................................................................... 5600/9188
2019-10-27T21:18:09.0413029Z .................................................................................................... 5700/9188
2019-10-27T21:18:09.0413029Z .................................................................................................... 5700/9188
2019-10-27T21:18:18.5359701Z .....................................................ii...i..ii...........i......................... 5800/9188
2019-10-27T21:18:44.4089476Z .................................................................................................... 6000/9188
2019-10-27T21:18:53.8962740Z .................................................................................................... 6100/9188
2019-10-27T21:18:53.8962740Z .................................................................................................... 6100/9188
2019-10-27T21:18:59.1408937Z ............................................................................i..ii................... 6200/9188
2019-10-27T21:19:29.4976759Z .................................................................................................... 6400/9188
2019-10-27T21:19:31.8710606Z ..........................................i......................................................... 6500/9188
2019-10-27T21:19:34.1883773Z .................................................................................................... 6600/9188
2019-10-27T21:19:36.7030964Z .................i.................................................................................. 6700/9188
---
2019-10-27T21:24:14.3840451Z  finished in 5.724
2019-10-27T21:24:14.4023085Z Check compiletest suite=codegen mode=codegen (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-10-27T21:24:14.5613178Z 
2019-10-27T21:24:14.5614313Z running 153 tests
2019-10-27T21:24:17.8174186Z i....iii......iii..iiii...i.............................i..i..................i....i...........ii.i. 100/153
2019-10-27T21:24:19.8084244Z i..iiii..............i.........iii.i.........ii......
2019-10-27T21:24:20.7376021Z 
2019-10-27T21:24:20.7420248Z  finished in 5.406
2019-10-27T21:24:20.7421399Z Check compiletest suite=codegen-units mode=codegen-units (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-10-27T21:24:20.7421643Z 
---
2019-10-27T21:24:22.0119961Z  finished in 2.184
2019-10-27T21:24:22.0308757Z Check compiletest suite=assembly mode=assembly (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-10-27T21:24:22.1919428Z 
2019-10-27T21:24:22.1920193Z running 9 tests
2019-10-27T21:24:22.1921129Z iiiiiiiii
2019-10-27T21:24:22.2128408Z 
2019-10-27T21:24:22.2128597Z  finished in 0.161
2019-10-27T21:24:22.2129311Z Check compiletest suite=incremental mode=incremental (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-10-27T21:24:22.7308727Z 
---
2019-10-27T21:24:40.9058058Z  finished in 18.693
2019-10-27T21:24:40.9266630Z Check compiletest suite=debuginfo mode=debuginfo-gdb+lldb (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-10-27T21:24:41.0941075Z 
2019-10-27T21:24:41.0941391Z running 123 tests
2019-10-27T21:25:06.2461931Z .iiiii...i.....i..i...i..i.i.i..i.ii..i.i.....i..i....ii..........iiii..........i...ii...i.......ii. 100/123
2019-10-27T21:25:11.1279885Z i.i.i......iii.i.....ii
2019-10-27T21:25:11.1280500Z 
2019-10-27T21:25:11.1280573Z  finished in 30.201
2019-10-27T21:25:11.1291507Z Uplifting stage1 rustc (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-10-27T21:25:11.1292004Z Copying stage2 rustc from stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
---
2019-10-27T21:37:50.8576704Z 
2019-10-27T21:37:50.8577655Z    Doc-tests core
2019-10-27T21:37:56.2118469Z 
2019-10-27T21:37:56.2119392Z running 2409 tests
2019-10-27T21:38:07.5159311Z ......iiiii......................................................................................... 100/2409
2019-10-27T21:38:18.5768890Z ................................................................................ii..F............... 200/2409
2019-10-27T21:38:44.5941132Z ..i................................................................................................. 400/2409
2019-10-27T21:38:44.5941132Z ..i................................................................................................. 400/2409
2019-10-27T21:38:55.2654295Z .................................................i..i.................iiii.......................... 500/2409
2019-10-27T21:39:16.9582733Z .................................................................................................... 700/2409
2019-10-27T21:39:27.0643009Z .................................................................................................... 800/2409
2019-10-27T21:39:37.5212318Z .................................................................................................... 900/2409
2019-10-27T21:39:47.9921926Z .................................................................................................... 1000/2409
---
2019-10-27T21:42:24.3394857Z 
2019-10-27T21:42:24.3437085Z error: test failed, to rerun pass '--doc'
2019-10-27T21:42:24.3454344Z 
2019-10-27T21:42:24.3454416Z 
2019-10-27T21:42:24.3464610Z command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "-Zconfig-profile" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "2" "--release" "--locked" "--color" "always" "--features" "panic-unwind backtrace compiler-builtins-c" "--manifest-path" "/checkout/src/libtest/Cargo.toml" "-p" "core" "--" "--quiet"
2019-10-27T21:42:24.3464760Z 
2019-10-27T21:42:24.3464810Z 
2019-10-27T21:42:24.3470255Z failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
2019-10-27T21:42:24.3499360Z Build completed unsuccessfully in 1:24:14
2019-10-27T21:42:24.3499360Z Build completed unsuccessfully in 1:24:14
2019-10-27T21:42:24.3528351Z == clock drift check ==
2019-10-27T21:42:24.3565703Z   local time: Sun Oct 27 21:42:24 UTC 2019
2019-10-27T21:42:27.7731263Z   network time: Sun, 27 Oct 2019 21:42:27 GMT
2019-10-27T21:42:27.7739331Z == end clock drift check ==
2019-10-27T21:42:28.2560012Z 
2019-10-27T21:42:28.2624561Z ##[error]Bash exited with code '1'.
2019-10-27T21:42:28.2666104Z ##[section]Starting: Checkout
2019-10-27T21:42:28.2668121Z ==============================================================================
2019-10-27T21:42:28.2668174Z Task         : Get sources
2019-10-27T21:42:28.2668236Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:umbrella: The latest upstream changes (presumably #65188) made this pull request unmergeable. Please resolve the merge conflicts.

ohadravid commented:
Apparently the example used in the [documentation](https://github.com/rust-lang/rust/blob/master/src/libcore/convert.rs#L227) for the `Into` trait fails because it can now be complied.

This is the expected behavior (and one of the RFC's stated goals), but I don't know what to do with it (remove the example? change it to say that `before rust. 1.40...`?).

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=12112&view=logs&jobId=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/12112/logs/49), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/12112/logs/49)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-10-28T12:03:44.8101181Z ##[command]git remote add origin https://github.com/rust-lang/rust
2019-10-28T12:03:44.8318506Z ##[command]git config gc.auto 0
2019-10-28T12:03:44.8399239Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2019-10-28T12:03:44.8465054Z ##[command]git config --get-all http.proxy
2019-10-28T12:03:44.8625131Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/65879/merge:refs/remotes/pull/65879/merge
---
2019-10-28T13:02:37.8383904Z .................................................................................................... 1500/9190
2019-10-28T13:02:43.9608386Z .................................................................................................... 1600/9190
2019-10-28T13:02:54.1281439Z ..........................................................................................i......... 1700/9190
2019-10-28T13:03:01.2537219Z ......i............................................................................................. 1800/9190
2019-10-28T13:03:08.1562253Z ................................................................................iiiii............... 1900/9190
2019-10-28T13:03:29.7200727Z .................................................................................................... 2100/9190
2019-10-28T13:03:32.1400683Z .................................................................................................... 2200/9190
2019-10-28T13:03:34.8032310Z .................................................................................................... 2300/9190
2019-10-28T13:03:49.0261368Z .................................................................................................... 2400/9190
---
2019-10-28T13:06:46.6551442Z ................................................................................i...............i... 4700/9190
2019-10-28T13:06:54.3250536Z .................................................................................................... 4800/9190
2019-10-28T13:07:04.4787858Z .................................................................................................... 4900/9190
2019-10-28T13:07:09.9271141Z .................................................................................................... 5000/9190
2019-10-28T13:07:21.7440705Z .................................................................................ii.ii...........i.. 5100/9190
2019-10-28T13:07:30.5852670Z .................................................................................................... 5300/9190
2019-10-28T13:07:40.1212924Z .................................................................................................... 5400/9190
2019-10-28T13:07:46.7258765Z ...................................................i................................................ 5500/9190
2019-10-28T13:07:53.5602709Z .................................................................................................... 5600/9190
2019-10-28T13:07:53.5602709Z .................................................................................................... 5600/9190
2019-10-28T13:08:04.0726223Z .................................................................................................... 5700/9190
2019-10-28T13:08:14.8626143Z ....................................ii...i..ii...........i.......................................... 5800/9190
2019-10-28T13:08:37.4421642Z .................................................................................................... 6000/9190
2019-10-28T13:08:43.8114888Z .................................................................................................... 6100/9190
2019-10-28T13:08:43.8114888Z .................................................................................................... 6100/9190
2019-10-28T13:08:50.9642083Z .......................................................i..ii........................................ 6200/9190
2019-10-28T13:09:18.4209270Z .................................................................................................... 6400/9190
2019-10-28T13:09:20.5400057Z .....................i.............................................................................. 6500/9190
2019-10-28T13:09:22.7749207Z ................................................................................................i... 6600/9190
2019-10-28T13:09:25.4367279Z .................................................................................................... 6700/9190
---
2019-10-28T13:13:53.5426753Z  finished in 5.490
2019-10-28T13:13:53.5603416Z Check compiletest suite=codegen mode=codegen (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-10-28T13:13:53.7540337Z 
2019-10-28T13:13:53.7541913Z running 153 tests
2019-10-28T13:13:56.8869589Z i....iii......iii..iiii...i.............................i..i..................i....i...........ii.i. 100/153
2019-10-28T13:13:58.8196334Z i..iiii..............i.........iii.i.........ii......
2019-10-28T13:13:58.8196992Z 
2019-10-28T13:13:58.8201184Z  finished in 5.260
2019-10-28T13:13:58.8388865Z Check compiletest suite=codegen-units mode=codegen-units (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-10-28T13:13:59.0274682Z 
---
2019-10-28T13:14:00.9897086Z  finished in 2.150
2019-10-28T13:14:01.0067243Z Check compiletest suite=assembly mode=assembly (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-10-28T13:14:01.1632482Z 
2019-10-28T13:14:01.1632721Z running 9 tests
2019-10-28T13:14:01.1637187Z iiiiiiiii
2019-10-28T13:14:01.1638197Z 
2019-10-28T13:14:01.1642013Z  finished in 0.157
2019-10-28T13:14:01.1839355Z Check compiletest suite=incremental mode=incremental (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-10-28T13:14:01.3791339Z 
---
2019-10-28T13:14:19.4485196Z  finished in 18.264
2019-10-28T13:14:19.4731516Z Check compiletest suite=debuginfo mode=debuginfo-gdb+lldb (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-10-28T13:14:19.6604673Z 
2019-10-28T13:14:19.6605014Z running 123 tests
2019-10-28T13:14:43.8866295Z .iiiii...i.....i..i...i..i.i.i..i.ii..i.i.....i..i....ii..........iiii..........i...ii...i.......ii. 100/123
2019-10-28T13:14:48.5884870Z i.i.i......iii.i.....ii
2019-10-28T13:14:48.5885416Z 
2019-10-28T13:14:48.5892018Z  finished in 29.116
2019-10-28T13:14:48.5901214Z Uplifting stage1 rustc (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-10-28T13:14:48.5901629Z Copying stage2 rustc from stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
---
2019-10-28T13:27:04.1168987Z 
2019-10-28T13:27:04.1169774Z    Doc-tests core
2019-10-28T13:27:09.1848511Z 
2019-10-28T13:27:09.1849509Z running 2409 tests
2019-10-28T13:27:20.0451537Z ......iiiii......................................................................................... 100/2409
2019-10-28T13:27:30.7642250Z ................................................................................ii.F................ 200/2409
2019-10-28T13:27:55.8872181Z ..i................................................................................................. 400/2409
2019-10-28T13:27:55.8872181Z ..i................................................................................................. 400/2409
2019-10-28T13:28:06.3368471Z .................................................i..i.................iiii.......................... 500/2409
2019-10-28T13:28:26.1449795Z .................................................................................................... 700/2409
2019-10-28T13:28:36.3052128Z .................................................................................................... 800/2409
2019-10-28T13:28:46.5886725Z .................................................................................................... 900/2409
2019-10-28T13:28:56.8213779Z .................................................................................................... 1000/2409
---
2019-10-28T13:31:28.1985695Z 
2019-10-28T13:31:28.2022848Z error: test failed, to rerun pass '--doc'
2019-10-28T13:31:28.2044432Z 
2019-10-28T13:31:28.2045849Z 
2019-10-28T13:31:28.2048326Z command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "-Zconfig-profile" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "2" "--release" "--locked" "--color" "always" "--features" "panic-unwind backtrace compiler-builtins-c" "--manifest-path" "/checkout/src/libtest/Cargo.toml" "-p" "core" "--" "--quiet"
2019-10-28T13:31:28.2050206Z 
2019-10-28T13:31:28.2051201Z 
2019-10-28T13:31:28.2062610Z failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
2019-10-28T13:31:28.2064038Z Build completed unsuccessfully in 1:21:19
2019-10-28T13:31:28.2064038Z Build completed unsuccessfully in 1:21:19
2019-10-28T13:31:28.2115635Z == clock drift check ==
2019-10-28T13:31:28.2132442Z   local time: Mon Oct 28 13:31:28 UTC 2019
2019-10-28T13:31:28.3569342Z   network time: Mon, 28 Oct 2019 13:31:28 GMT
2019-10-28T13:31:28.3572103Z == end clock drift check ==
2019-10-28T13:31:28.9840642Z 
2019-10-28T13:31:28.9987372Z ##[error]Bash exited with code '1'.
2019-10-28T13:31:29.0021888Z ##[section]Starting: Checkout
2019-10-28T13:31:29.0023900Z ==============================================================================
2019-10-28T13:31:29.0023956Z Task         : Get sources
2019-10-28T13:31:29.0024005Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:umbrella: The latest upstream changes (presumably #65919) made this pull request unmergeable. Please resolve the merge conflicts.

nikomatsakis commented:
@ohadravid 

> Apparently the example used in the documentation for the Into trait fails because it can now be complied.

amusing

nikomatsakis commented:
I wonder if this means that whole section can be removed, and we can just advice to *always* implement `From`?

nikomatsakis commented on src/libcore/convert.rs:
```diff
@@ -220,21 +220,20 @@ pub trait AsMut<T: ?Sized> {
 ///
-/// If the destination type is not part of the current crate
-/// then you can't implement [`From`] directly.
+/// Prior to Rust 1.40, if the destination type was not part of the current crate
```
I think we should strengthen this -- unless we have a good example of when you *do* need to implement `Into`, we should probably start by telling people to **avoid** implementing `Into` directly, and then continue with this content?

dalcde replied:
This should say 1.41 instead?

ohadravid commented:
As far as I can tell, there is no reason to implement `Into`, so I changed it to be more forceful.
However I just might not be able to come up with a contrived-enough example.

(I went to `diesel` and changed all the `Into` to `From` and it worked where it didn't before `re_rebalance`, so that's something)

ohadravid commented:
@nikomatsakis I changed the docs like you suggested! Let me know if there's anything else to do 😁

Centril commented on src/libsyntax/feature_gate/accepted.rs:
```diff
@@ -253,6 +253,9 @@ declare_features! (
     (accepted, cfg_doctest, "1.40.0", Some(62210), None),
+    /// Allows relaxing the coherence rules such that
+    /// `impl<T> ForeignTrait<LocalType> for ForeignType<T>` is permitted.
+    (accepted, re_rebalance_coherence, "1.40.0", Some(55437), None),
```
```suggestion
    (accepted, re_rebalance_coherence, "1.41.0", Some(55437), None),
```

nikomatsakis commented:
@bors r+

bors commented:
:pushpin: Commit 3e0759dc0565c3b8e5ca0ab3ace15ae7e91b6ffd has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 3e0759dc0565c3b8e5ca0ab3ace15ae7e91b6ffd -->
<!-- homu: {"type":"Approved","sha":"3e0759dc0565c3b8e5ca0ab3ace15ae7e91b6ffd","approver":"nikomatsakis"} -->

Centril commented:
@bors rollup=never p=3

bors commented:
:hourglass: Testing commit 3e0759dc0565c3b8e5ca0ab3ace15ae7e91b6ffd with merge 98c173afe4b7019d0b739151b7d4d4343a85ea2b...
<!-- homu: {"type":"BuildStarted","head_sha":"3e0759dc0565c3b8e5ca0ab3ace15ae7e91b6ffd","merge_sha":"98c173afe4b7019d0b739151b7d4d4343a85ea2b"} -->

bors commented:
:sunny: Test successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=13077)
Approved by: nikomatsakis
Pushing 98c173afe4b7019d0b739151b7d4d4343a85ea2b to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nikomatsakis","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=13077"},"merge_sha":"98c173afe4b7019d0b739151b7d4d4343a85ea2b"} -->
